contributors/design-proposals/sig-storage/volume-topology-scheduling.md

```diff
@@ -0,0 +1,580 @@
+
+##### Pull entire PVC binding into the scheduler
+The proposed design only has the scheduler initiating the binding transaction
+by prebinding the PV.  An alternative is to pull the whole two-way binding
+transaction into the scheduler, but there are some complex scenarios that
+scheduler’s Pod sync loop cannot handle:
+* PVC and PV getting unexpectedly unbound or lost
+* PVC and PV state getting partially updated
+* PVC and PV deletion and cleanup
+
+Handling these scenarios in the scheduler’s Pod sync loop is not possible, so
+they have to remain in the PV controller.
+
+##### Keep all PVC binding in the PV controller
+Add a new “node” annotation to the PVC object and have the scheduler update it
+when it schedules the pod to a node.  A new scheduling predicate is still
+needed to filter and match the PVs (but not actually bind).
+
+Major downsides of this approach include:
+* Requires PV controller to change its sync loop to operate on pods, in order
+to handle the multiple PVCs in a pod scenario.  This is a potentially big change
+that would be hard to keep separate and feature-gated from the current PV logic.
+* Needs new coordination between PV controller and kubelet to reject a pod if
+binding fails to send the pod back to scheduler.
+* Initial filtering (matching) and binding are done asynchronously so there is a
+higher chance of binding failures.  As an example, we could match PV1 for both
+pod1 and pod2, but only pod1 will be able to bind successfully, and pod2 has to
+be sent back to scheduling.
+* More race conditions and higher chance of binding failure because of 3
+different components operating independently (scheduler, PV controller, kubelet).
+
+##### Move PVC binding to kubelet
+Looking into the future, with the potential for NUMA-aware scheduling, you could
+have a sub-scheduler on each node to handle the pod scheduling within a node.  It
+could make sense to have the volume binding as part of this sub-scheduler, to make
+sure that the volume selected will have NUMA affinity with the rest of the
+resources that the pod requested.
+
+However, there are potential security concerns because kubelet would need to see
+unbound PVs in order to bind them.  For local storage, the PVs could be restricted
+to just that node, but for zonal storage, it could see all the PVs in that zone.
+
+In addition, the sub-scheduler is just a thought at this point, and there are no
+concrete proposals in this area yet.
+
+### Binding multiple PVCs in one transaction
+For the alpha phase, this is not required.  Since the scheduler is serialized, a
+partial binding failure should be a rare occurrence.  Manual recovery will need
+to be done for those rare failures.
+
+### Recovering from kubelet rejection of pod
+Again, for alpha phase, manual intervention will be required.  For beta, we’ll
+have to consider some sort of rollback mechanism for more robustness.  Perhaps
+this rollback mechanism can also be used for binding multiple PVCs in one
```

Comment:
I think that recycler (and deleter) is safe here, PV controller should be robust enough already.

On the other hand, any rollback in recycler may conflict with binding in PV controller. Scheduler sets pv.spec.claimRef and leaves the rest to PV controller, right? PV controller will save pvc.spec.volumeName, update status and some annotations. In case the scheduler wants to revert the binding, clearing pv.spec.claimRef won't be enough, it needs to clear everything. And it will race with the PV controller trying to fix the binding, it tries to fix the status and binding if something overwrites it.

Reply:
Yah we will need to add some new state so that the PV controller will not try to fix the binding on a rollback.

Reply:
edit: rollback in recycler -> rollback in scheduler